client/src/main/scala/com.gu.contentapi.client/ContentApiClient.scala

```diff
@@ -106,8 +106,36 @@ trait ContentApiClient {
     paginate(query)(f).map {
       case Nil => throw new RuntimeException("Something went wrong with the query")
       case m :: Nil => m
-      case m1 :: m2 :: ms => ms.foldLeft(g(m1, m2))(g)
+      case ms => ms.reduce(g)
+    }
+
+  /** Unfolds a query by accumulating its results
+    * 
+    * @tparam Q the type of a Content API query with pagination parameters
+    * @tparam R the type of response expected for `Q`
+    * @param query the initial query
+    * @param f a result-processing function
+    * @return a future of an accumulated value
+    */
+  def paginateFold[Q <: PaginatedApiQuery[Q], R, M](query: Q)(f: (R, M) => M, m: M)(
```

Comment:
For consistency with the Scala library API, you may want to change the order of parameters to accept the starting value in a separate argument list. It can also help with type-inferencing.
```scala
def paginateFold[Q <: PaginatedApiQuery[Q], R, M](query: Q)(m: M)(f: (R, M) => M)(implicit ...)
// can be used as:
paginateFold(query)(0) { (response, sum) => sum + response.results.length }
```

Reply:
Sadly, it does not work. This has been bugging me a lot 🌋  The compiler complains that the type of `response` is missing. If you know how to get around that, then it's 🌈

Reply:
I see. It looks like there is no connection between `SearchQuery` and `SearchResponse`. (Or, at least, I can't find one that the compiler can use).

I haven't tried to compile it but it looks like it's possible to call this as `api.paginateFold(ContentApiClient.search) { (_: ItemResponse, _) => }`

Reply:
It would probably fail with `could not find implicit value for parameter decoder: Decoder.Aux[SearchQuery, ItemResponse]`. I think I overestimated the compiler, you could actually have multiple instances of `Decoder.Aux[T, Ri]` with different `Ri`s.